### PR TITLE
Make scheduler-plugins the default gang scheduler

### PR DESCRIPTION
**Pull request by Syulin7**
**What this PR does / why we need it:**

Now supports many gang schedulers(volcano, scheduler-plugins). This PR supports koordinator gang scheduler. 

Related: https://github.com/kubeflow/training-operator/issues/1746

### CODE REVIEW

**Comment by johnugeorge** [comment]
Thanks @Syulin7 We are in the process of training operator release(currently rc0).  Since kubeflow/common dependency has not branched out, we have to wait a bit to have the final release and then merge this one. 


/cc @tenzen-y  Can you review this?

**Comment by tenzen-y** [comment]
> Thanks @Syulin7 We are in the process of training operator release(currently rc0). Since kubeflow/common dependency has not branched out, we have to wait a bit to have the final release and then merge this one.
> 
> /cc @tenzen-y Can you review this?

Sure. 
First, I will ask questions about the Koordinator at https://github.com/kubeflow/training-operator/issues/1746.

File: pkg/controller.v1/control/podgroup_control.go
Comment (tenzen-y) [inline pkg/controller.v1/control/podgroup_control.go:269]:
```suggestion
// KoordinatorControl is the  implementation of PodGroupControlInterface with scheduler-plugins.
type KoordinatorControl struct {
	*SchedulerPluginsControl
}

var _ PodGroupControlInterface = &KoordinatorControl{}

func (k *KoordinatorControl) DecoratePodTemplateSpec(pts *corev1.PodTemplateSpec, job metav1.Object, _ string) {
	if len(pts.Spec.SchedulerName) == 0 {
		pts.Spec.SchedulerName = k.GetSchedulerName()
	}
	if pts.Labels == nil {
		pts.Labels = make(map[string]string)
	}
	pts.Labels[schedulerpluginsv1alpha1.PodGroupLabel] = job.GetName()
}

func (k *KoordinatorControl) GetSchedulerName() string {
	return "koord-scheduler"
}
```

File: pkg/reconciler.v1/common/gang_koordinator.go
Comment (tenzen-y) [inline pkg/reconciler.v1/common/gang_koordinator.go:195]:
```suggestion
type KoordinatorReconciler struct {
	*SchedulerFrameworkReconciler
}

func BareKoordinatorReconciler(client client.Client, bgReconciler *BaseGangReconciler, enabled bool) *KoordinatorReconciler {
	if bgReconciler == nil {
		bgReconciler = &BaseGangReconciler{}
	}
	bgReconciler.Enabled = enabled
	return &KoordinatorReconciler{SchedulerFrameworkReconciler: BareSchedulerFrameworkReconciler(client, bgReconciler, enabled)}
}
```

Comment (Syulin7) [reply pkg/reconciler.v1/common/gang_koordinator.go]: Done, thanks for your suggestion.

**Comment by Syulin7** [comment]
@tenzen-y This PR is ready for review. PTAL.

File: pkg/controller.v1/common/job_controller.go
Comment (tenzen-y) [inline pkg/controller.v1/common/job_controller.go:201]: Why are there the same functions (L181-L187)?

**Comment by tenzen-y** [comment]
Can you squash commits into one?

**Comment by Syulin7** [comment]
> Can you squash commits into one?

Done

File: pkg/controller.v1/common/job_controller.go
Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:194]: Done

**Review by tenzen-y — COMMENTED** [review]
/lgtm
/assign @johnugeorge

**Comment by johnugeorge** [comment]
@Syulin7 We will get back to this in a week. Sorry for the delay as we haven't released 1.6 release yet.

**Comment by alculquicondor** [comment]
IMO, this is not sustainable. We'll keep adding support for more schedulers in the ecosystem?

schedulers need to look into generic approaches, such as the suspend field.

**Comment by tenzen-y** [comment]
> IMO, this is not sustainable. We'll keep adding support for more schedulers in the ecosystem?
> 
> schedulers need to look into generic approaches, such as the suspend field.

@alculquicondor We don't have any criteria to add more schedulers. So we may need to have the criteria (e.g., only CNCF project?)

**Comment by alculquicondor** [comment]
An alternative that I'm proposing in kubeflow/mpi-operator#500 is that koordinator implements this on a controller in their repo.

I don't see why that's not possible.

**Comment by tenzen-y** [comment]
> An alternative that I'm proposing in [kubeflow/mpi-operator#500](https://github.com/kubeflow/mpi-operator/issues/500) is that koordinator implements this on a controller in their repo.
> 
> I don't see why that's not possible.

It looks like you have a point.
However, this issue applies to volcano and scheduler-plugins as well, not just to the koordinator.

**Comment by alculquicondor** [comment]
If this works out well, it could be a precedent for justifying https://github.com/kubernetes/kubernetes/issues/107294, which all schedulers can use to generically group any object.

**Comment by tenzen-y** [comment]
> If this works out well, we could justify:
> 
> 1. Adding similar fields to kubeflow's `schedulingPolicy` in `batch/v1.Job`
> 2. [Add suspend subresource kubernetes/kubernetes#107294](https://github.com/kubernetes/kubernetes/issues/107294), which all schedulers can use to generically group pods from a custom job.

Totally, I agree. However, since we may need long terms to complete all steps, I would like to add coscheduling support to MPIJob for the short term.

**Comment by tenzen-y** [comment]
> Yes, it applies to all.
> 
> If the maintainers of kubeflow agree, we would have to:
> 
> 1. Open issues in volcano and scheduling-plugins to implement the logic in their side.
> 2. Make it possible to disable the PodGroup control in kubeflow (I believe this is possible by just leaving the `gang-scheduler` parameter empty).
> 3. Remove the code after a few releases.

Ah, I missed the above comments. I'm ok with having the logic for PodGroup in the scheduler-plugins repo.
If the scheduler-plugins team agrees with this, I can help to implement the logic on their repo.

**Comment by alculquicondor** [comment]
For the time being, I would vote for only keeping support for the schedulers we already have (scheduler-plugins and volcano).

I would encourage the maintainers of Koordinator to put their efforts in https://github.com/kubernetes/kubernetes/issues/107294 and https://github.com/kubernetes/enhancements/issues/3370 instead of adding more dependencies in kubeflow.

**Comment by Syulin7** [comment]
@alculquicondor  I understand your concerns, and it make sense. As I mentioned in https://github.com/kubeflow/training-operator/issues/1746, the Koordinator scheduler uses the PodGroup defined in scheduler-plugins/coscheduling. Therefore, supporting Koordinator in the training-operator will not add more dependencies, but rather provide another option. Currently, some people are using both the training-operator and Koordinator simultaneously, which is very useful for them.

In the long run, I think it's necessary to provide a Native PodGroup API that includes fields such as suspend, queue, minNumOfWorkers, etc. Different schedulers can implement the same Native PodGroup API. For example, the operator only needs to add a specific label to the pod, and the scheduler creates and maintains the PodGroup. This way, the operator and scheduler won't depend on each other. I think we can work towards this direction in the future, but supporting Koordinator in the training-operator currently will not affect this.

@alculquicondor @tenzen-y @terrytangyuan  WDYT

**Comment by tenzen-y** [comment]
> Therefore, supporting Koordinator in the training-operator will not add more dependencies, but rather provide another option. 

That's right for now. However, the training-operator must directly depend on the koordinator if the koordinator-scheduler has original logic or CRDs in the future.

However, fortunately, the users can deploy CustomJobs with koordinator-scheduler by specifying scheduler-name even if we don't add this feature to the master branch as I mentioned in https://github.com/kubeflow/training-operator/issues/1746#issuecomment-1408364031.

So I agree with @alculquicondor.

File: pkg/controller.v1/common/job.go
Comment (alculquicondor) [inline pkg/controller.v1/common/job.go:278]: why is the PR bigger than just this line?

if we can have this, I would support this change. Otherwise, I'm opposed.

```golang
switch jc.Config.GangScheduling {
case GangSchedulerVolcano:
  ...
default:
}
```

File: pkg/controller.v1/control/podgroup_control.go
Comment (tenzen-y) [inline pkg/controller.v1/control/podgroup_control.go:210]: > why is the PR bigger than just this line?
> 
> if we can have this, I would support this change. Otherwise, I'm opposed.

@alculquicondor We need to add this function if we natively support the koordinator-scheduler.
So, you may oppose this change.

File: pkg/controller.v1/common/job.go
Comment (alculquicondor) [reply pkg/controller.v1/common/job.go:278]: here, I'm considering scheduling-plugins the "default" because it has a higher chance of becoming part of upstream kubernetes at some point.

File: pkg/controller.v1/control/podgroup_control.go
Comment (alculquicondor) [reply pkg/controller.v1/control/podgroup_control.go:210]: This is no different from SchedulerPluginsControl. They can remove `KoordinatorControl` and use the existing implementation.

Comment (tenzen-y) [reply pkg/controller.v1/control/podgroup_control.go:210]: Sorry for your confusion.
We must handle another value in the function GetSchedulerName (L212 - L214 in this file).

This means we must specify the `koord-scheduler` to the schedulerName in podSpec.

Comment (alculquicondor) [reply pkg/controller.v1/control/podgroup_control.go:210]: if you treat `SchedulerPluginsControl` as the default implementation when the gang-scheduler-plugin configuration is set, you can pass that.

Comment (Syulin7) [reply pkg/controller.v1/control/podgroup_control.go:210]: > if you treat `SchedulerPluginsControl` as the default implementation when the gang-scheduler-plugin configuration is set, you can pass that string, even if it's not "scheduler-plugins"

I think this is ok,add --enable-gang-scheduler and --gang-scheduler-name. SchedulerPluginsControl as the default implementation when the enable-gang-scheduler is true, and set gang-scheduler-name if it is not nil.

Comment (alculquicondor) [reply pkg/controller.v1/control/podgroup_control.go:210]: just not empty gang-scheduler-name should be enough. No need for an extra enable-gang-scheduler

Comment (alculquicondor) [reply pkg/controller.v1/control/podgroup_control.go:210]: sgtm

Comment (tenzen-y) [reply pkg/controller.v1/control/podgroup_control.go:210]: > scheduler-plugins as secondary scheduler, set gang-scheduler-name=scheduler-plugins(or any other name they use. like koordinator.) scheduler-plugins as default scheduler, set gang-scheduler-name=default-scheduler volcano as secondary scheduler, set gang-scheduler-name=volcano @tenzen-y @alculquicondor WDYT

@Syulin7 I have a question. If we set gang-scheduler-name=default-scheduler, does the training-operator embed `default-scheduler` in podTemplateSpec.spec.schedulerName? Or does the training-operator skip set a value in podTemplateSpec.spec.schedulerName?

Comment (Syulin7) [reply pkg/controller.v1/control/podgroup_control.go:210]: @tenzen-y  I think if we set gang-scheduler-name=default-scheduler, the training-operator should skip set a value in podTemplateSpec.spec.schedulerName.

Comment (tenzen-y) [reply pkg/controller.v1/control/podgroup_control.go:210]: sgtm

**Comment by Syulin7** [comment]
@tenzen-y @alculquicondor  This PR is ready for review. PTAL.

I set SchedulerPluginsControl as the default implementation. 
Removed koordinator related code, set --gang-scheduler-name=koord-scheduler can support Koordinator Gang Scheduler.

The test results are as follows.
```
$ kubectl get pods -n kubeflow
NAME                    READY   STATUS    RESTARTS   AGE
tfjob-simple-worker-0   0/1     Pending   0          82s
tfjob-simple-worker-1   0/1     Pending   0          82s
$ kubectl describe pods -n kubeflow tfjob-simple-worker-0 | grep -A10  Events
Events:
  Type     Reason            Age   From             Message
  ----     ------            ----  ----             -------
  Warning  FailedScheduling  16m   koord-scheduler  running PreFilter plugin "Coscheduling": gang child pod not collect enough, gangName: kubeflow/tfjob-simple, podName: kubeflow/tfjob-simple-worker-0
  Warning  FailedScheduling  16m   koord-scheduler  running PreFilter plugin "Coscheduling": gang child pod not collect enough, gangName: kubeflow/tfjob-simple, podName: kubeflow/tfjob-simple-worker-0
```

File: pkg/controller.v1/common/job_controller.go
Comment (tenzen-y) [inline pkg/controller.v1/common/job_controller.go:182]: Should we enforce an empty string? 

```go
jc.Config.GangScheduling = ""
```

File: pkg/controller.v1/control/podgroup_control.go
Comment (tenzen-y) [inline pkg/controller.v1/control/podgroup_control.go:195]: Why is this removed?

File: pkg/controller.v1/common/job_controller.go
Comment (alculquicondor) [inline pkg/controller.v1/common/job_controller.go:74]: Add a comment that using this scheduler name or any scheduler name different than volcano uses the scheduler-plugins PodGroup

File: pkg/controller.v1/common/job_controller.go
Comment (alculquicondor) [inline pkg/controller.v1/common/job_controller.go:72]: oh I wasn't aware that this existed. Maybe we can add another:

```
case GangSchedulerNone, "":
  return nil
default:
  // scheduler plugins
```

File: pkg/controller.v1/common/pod.go
Comment (alculquicondor) [inline pkg/controller.v1/common/pod.go:424]: was this assuming that the default scheduler was volcano? Or is the comment innacurate?

File: pkg/controller.v1/control/podgroup_control.go
Comment (alculquicondor) [inline pkg/controller.v1/control/podgroup_control.go:133]: why check `GetSchedulerName != "default-schedule"`?

File: pkg/reconciler.v1/common/gang_scheduler_framework.go
Comment (alculquicondor) [inline pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: What is this for @tenzen-y ?

Comment (tenzen-y) [reply pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: Maybe, this function is expected using in tests like the following:

https://github.com/kubeflow/common/blob/4dbf18c85417b54158914d1d01c1744b4e3542f2/test_job/reconciler.v1/test_job/test_job_reconciler.go#L55

Comment (tenzen-y) [reply pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: However, we don't have any test reconciler for scheduler-plugins for now.

Comment (alculquicondor) [reply pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: so it basically schedules the pods?

Comment (tenzen-y) [reply pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: Ah, I remind the purpose of this function.
This function isn't only for tests.

This just creates a new instance for the `SchedulerFrameworkReconciler`.

However, we use controllers instead of reconcilers. So we don't use this package for now.
More details:  #140

File: pkg/controller.v1/common/pod.go
Comment (tenzen-y) [reply pkg/controller.v1/common/pod.go:424]: > was this assuming that the default scheduler was volcano?

We used to decorate the schedulerName in podTemplateSpec.spec.schedulerName only when the gang-scheduler name is set as `volcano`.

- volcano

https://github.com/kubeflow/common/blob/4dbf18c85417b54158914d1d01c1744b4e3542f2/pkg/controller.v1/control/podgroup_control.go#L65-L74

- scheduler-plugins

https://github.com/kubeflow/common/blob/4dbf18c85417b54158914d1d01c1744b4e3542f2/pkg/controller.v1/control/podgroup_control.go#L131-L136

Comment (alculquicondor) [reply pkg/controller.v1/common/pod.go:424]: so the comment was inaccurate, right?

File: pkg/reconciler.v1/common/gang_scheduler_framework.go
Comment (Syulin7) [reply pkg/reconciler.v1/common/gang_scheduler_framework.go:44]: > However, we use controllers instead of reconcilers. So we don't use this package for now. More details: #140
yes, we don't use this package for now.

File: pkg/controller.v1/control/podgroup_control.go
Comment (Syulin7) [reply pkg/controller.v1/control/podgroup_control.go:195]: Done. My mistake

File: pkg/controller.v1/common/job_controller.go
Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:182]: Done

File: pkg/controller.v1/common/job_controller.go
Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:173]: it is used in training-operator: https://github.com/kubeflow/training-operator/pull/1747/files#diff-900bf2495e0ccd04b3bdbd8cfabea9ee5eb73a6e3982a0a928bd6ed6e4ebe2d6L124-L125

File: pkg/controller.v1/control/podgroup_control.go
Comment (Syulin7) [reply pkg/controller.v1/control/podgroup_control.go:133]: if gang-scheduler-name=default-scheduler, skip set a schedulerName in podTemplateSpec.spec.schedulerName.

if gang-scheduler-name is set to a different value(e.g. scheduler-plugins, koordinator), we should set the schedulerName in podTemplateSpec.spec.schedulerName.

File: pkg/controller.v1/common/job_controller.go
Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:72]: Done

File: pkg/controller.v1/common/job_controller.go
Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:74]: Done

File: pkg/controller.v1/common/pod.go
Comment (Syulin7) [reply pkg/controller.v1/common/pod.go:424]: I update the comment.

**Comment by Syulin7** [comment]
I resolved all conversations, PTAL, thanks. @tenzen-y @alculquicondor

File: pkg/controller.v1/common/job.go
Comment (tenzen-y) [inline pkg/controller.v1/common/job.go:279]: Probably, we don't need this condition since we confirm if the JobConfig has a config for gang scheduling in L232.

cc: @alculquicondor

File: pkg/controller.v1/common/job_controller.go
Comment (tenzen-y) [inline pkg/controller.v1/common/job_controller.go:72]: What intent of using an empty string instead of `None`?

Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:72]: if GangSchedulerNone is None, It may be considered as enabling the gang scheduling, but I think we can remove the GangSchedulerNone const.
```
var GenNonGangSchedulerSetupFunc = func() GangSchedulingSetupFunc {
	return func(jc *JobController) {
		jc.Config.GangScheduling = GangSchedulerNone
		jc.PodGroupControl = nil
	}
}

func (c *JobControllerConfiguration) EnableGangScheduling() bool {
	return c.GangScheduling != ""
}
```

File: pkg/controller.v1/common/job.go
Comment (Syulin7) [reply pkg/controller.v1/common/job.go:279]: Done

File: pkg/controller.v1/common/job_controller.go
Comment (tenzen-y) [reply pkg/controller.v1/common/job_controller.go]: Thanks for clarifying.

>  but I think we can remove the GangSchedulerNone const.

Sounds good. However, it might be better to do that after we merge kubeflow/common into kubeflow/training-operator since GangSchedulerNone might be worth it as documentation.

Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:72]: > since GangSchedulerNone might be worth it as documentation.

Make sense, let's keep it.

File: pkg/controller.v1/common/job_controller.go
Comment (alculquicondor) [inline pkg/controller.v1/common/job_controller.go:86]:
```suggestion
	return c.GangScheduling != "" and c.GangScheduling != GangSchedulerNone
```

File: pkg/controller.v1/control/podgroup_control.go
Comment (alculquicondor) [reply pkg/controller.v1/control/podgroup_control.go:133]: why not just always set?

File: pkg/controller.v1/common/job_controller.go
Comment (tenzen-y) [reply pkg/controller.v1/common/job_controller.go:72]: Yes, the "None" was already released on v0.4.6.

Comment (Syulin7) [reply pkg/controller.v1/common/job_controller.go:72]: Yes, we should keep it for backwards compatibility.

File: pkg/controller.v1/common/job_controller.go
Comment (alculquicondor) [inline pkg/controller.v1/common/job_controller.go:76]: Remove if this is not being used

**Review by alculquicondor — COMMENTED** [review]
a nit, otherwise lgtm

**Comment by alculquicondor** [comment]
/retitle Make scheduler-plugins the default gang scheduler

File: pkg/reconciler.v1/common/gang_scheduler_framework.go
Comment (tenzen-y) [inline pkg/reconciler.v1/common/gang_scheduler_framework.go:168]:
```suggestion
	if podTemplate.Spec.SchedulerName == "" {
```

Let's use a logic the same as the podgroup_control.

**Comment by tenzen-y** [comment]
I left a comment for nit. Overall, lgtm.

**Comment by Syulin7** [comment]
> I left a comment for nit. Overall, lgtm.

**Comment by tenzen-y** [comment]
@Syulin7 Thanks for the great work!
/lgtm

@johnugeorge 
Do you decide on the week for releasing the training operator v1.6, although releasing kubeflow v1.7 seems delayed?

> Week 23 - Distribution Testing Ends: Wednesday, Mar 22nd 2023

File: pkg/controller.v1/control/podgroup_control.go
Comment (alculquicondor) [inline pkg/controller.v1/control/podgroup_control.go:148]: Not about this PR, but I realized that some functions below are unnecessary, like `UpdatePodGroup`, and `CreatePodGroup`. They are casting unnecessarily: the controller-runtime client can already handle the casting via reflection. And I think `GetPodGroup` and `DeletePodGroup` could be refactored so that we don't need the casting either. Maybe even use generics.

**Review by terrytangyuan — APPROVED** [review]
Thanks!

/lgtm
/approve

**Comment by johnugeorge** [comment]
@tenzen-y The final release is not yet made due to delay in Kubeflow upstream. Sorry for the delay. This should not happen again in the future as we always branch out for all repos except common.  We will target this week end to merge- Friday 24 March

**Comment by tenzen-y** [comment]
> @tenzen-y The final release is not yet made due to delay in Kubeflow upstream. Sorry for the delay. This should not happen again in the future as we always branch out for all repos except common. We will target this week end to merge- Friday 24 March

Thanks for letting me know! I appreciate your hard work!

**Comment by johnugeorge** [comment]
Training operator release is made. This can be merged.

Thanks for the patience

**Comment by tenzen-y** [comment]
@Syulin7 Thank you!
/hold cancel

**Comment by tenzen-y** [comment]
@johnugeorge Can we cut a new release on this repo to move forward https://github.com/kubeflow/training-operator/pull/1747?

**Pull request merged**
